### PR TITLE
Extended key usage

### PR DESCRIPTION
**Pull request by cyphermox**
Weeeeee

### CODE REVIEW

File: MokManager.c [hunk 1]
```diff
@@ -406,7 +409,33 @@ static void show_x509_info (X509 *X509Cert, UINT8 *hash)
...
+			ASN1_OBJECT *obj = sk_ASN1_OBJECT_value(extusage, j);
+                        int buflen = 80;
+			char buf[buflen];
+
+			ZeroMem(&extkeyusage, sizeof(extkeyusage));
+
+			OBJ_obj2txt(buf, buflen, obj, 0);
+			CatPrint(&extkeyusage, L"OID: %a", buf);
+			text[i++] = StrDuplicate(extkeyusage.str);
+		}
```
Comment (cyphermox) [inline MokManager.c]: free extkeyusage before reusing

File: MokManager.c [hunk 2]
```diff
@@ -406,7 +409,33 @@ static void show_x509_info (X509 *X509Cert, UINT8 *hash)
...
+	text = AllocateZeroPool(sizeof(CHAR16 *) * (fields*3 + sk_ASN1_OBJECT_num(extusage)*3 + 1));
+
+	if (extusage) {
+		POOL_PRINT extkeyusage;
+		int j = 0;
+
+		ZeroMem(&extkeyusage, sizeof(serial_string));
+		text[i++] = StrDuplicate(L"[Extended Key Usage]");
+		CatPrint(&extkeyusage, L"num: %d", sk_ASN1_OBJECT_num(extusage));
+		text[i++] = StrDuplicate(extkeyusage.str);
```
Comment (cyphermox) [inline MokManager.c]: Probably should be removed.

File: MokManager.c [hunk 3]
```diff
@@ -406,7 +409,33 @@ static void show_x509_info (X509 *X509Cert, UINT8 *hash)
...
+
+		for (j = 0; j < sk_ASN1_OBJECT_num(extusage); j++) {
+			ASN1_OBJECT *obj = sk_ASN1_OBJECT_value(extusage, j);
+                        int buflen = 80;
```
Comment (cyphermox) [inline MokManager.c]: alignment issue

File: shim.c [hunk 4]
```diff
@@ -388,6 +392,34 @@ static BOOLEAN verify_x509(UINT8 *Cert, UINTN CertSize)
...
 }
 
+static BOOLEAN verify_eku(UINT8 *Cert, UINTN CertSize)
+{
+	X509 *x509;
+	CONST UINT8 *Temp = Cert;
+	EXTENDED_KEY_USAGE *eku;
+	ASN1_OBJECT *module_signing;
+
+	module_signing = OBJ_nid2obj(OBJ_create("1.3.6.1.4.1.2312.16.1.2",
```
Comment (cyphermox) [inline shim.c]: clean up objects at the end...

File: shim.c [hunk 5]
```diff
@@ -388,6 +392,34 @@ static BOOLEAN verify_x509(UINT8 *Cert, UINTN CertSize)
...
+	                                        NULL, NULL));
+
+	x509 = d2i_X509 (NULL, &Temp, (long) CertSize);
```
Comment (cyphermox) [inline shim.c]: Does this needs to be freed?

File: shim.c [hunk 6]
```diff
@@ -388,6 +392,34 @@ static BOOLEAN verify_x509(UINT8 *Cert, UINTN CertSize)
...
+	if (x509 != NULL) {
+		eku = X509_get_ext_d2i(x509, NID_ext_key_usage, NULL, NULL);
```
Comment (cyphermox) [inline shim.c]: same as above

File: MokManager.c [hunk 7]
```diff
@@ -406,7 +409,32 @@ static void show_x509_info (X509 *X509Cert, UINT8 *hash)
 		return;
 
 	i = 0;
-	text = AllocateZeroPool(sizeof(CHAR16 *) * (fields*3 + 1));
+
+	extusage = X509_get_ext_d2i(X509Cert, NID_ext_key_usage, NULL, NULL);
+	text = AllocateZeroPool(sizeof(CHAR16 *) * (fields*3 + sk_ASN1_OBJECT_num(extusage)*3 + 1));
```
Comment (tyhicks) [inline MokManager.c:414]: This seems to be allocating more memory than what is needed (not necessarily a bad thing).

For the new strings related to extended key usage, I think it should only be an additional `sk_ASN1_OBJECT_num(extusage) + 2` that is _only_ added when sk_ASN1_OBJECT_num(extusage) is greater than zero.

**Pull request closed without merging**